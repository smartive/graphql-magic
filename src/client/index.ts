// created from 'create-ts-index'

export * from './gql';
export * from './queries';
