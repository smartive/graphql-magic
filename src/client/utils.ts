import upperCase from 'lodash/upperCase';

export const constantCase = (str: string) => upperCase(str).replace(/ /g, '_');
