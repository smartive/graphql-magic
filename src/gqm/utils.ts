import { SourceFile, SyntaxKind, SyntaxList } from 'ts-morph';

export const findDeclarationInFile = (sourceFile: SourceFile, name: string) => {
  const syntaxList = sourceFile.getChildrenOfKind(SyntaxKind.SyntaxList)[0];
  if (!syntaxList) {
    throw new Error('No SyntaxList');
  }
  const declaration = findDeclaration(syntaxList, name);
  if (!declaration) {
    throw new Error('No rawModels declaration');
  }
  return declaration;
};

const findDeclaration = (syntaxList: SyntaxList, name: string) => {
  for (const variableStatement of syntaxList.getChildrenOfKind(SyntaxKind.VariableStatement)) {
    for (const declaration of variableStatement.getDeclarationList().getDeclarations()) {
      if (declaration.getName() === name) {
        return declaration;
      }
    }
  }
};
