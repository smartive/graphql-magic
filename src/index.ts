// created from 'create-ts-index'

export * from './api';
export * from './client';
export * from './db';
export * from './migrations';
export * from './models';
export * from './permissions';
export * from './resolvers';
export * from './schema';
export * from './utils';
export * from './context';
export * from './errors';
export * from './values';
