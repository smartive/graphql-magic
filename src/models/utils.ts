import { camelCase, lowerFirst, startCase } from 'lodash';
import {
  CustomField,
  EntityField,
  EntityModel,
  EnumField,
  EnumModel,
  InputModel,
  InterfaceModel,
  Model,
  ObjectModel,
  PrimitiveField,
  RawEnumModel,
  RelationField,
  ScalarModel,
} from './models';

const isNotFalsy = <T>(v: T | null | undefined | false): v is T => typeof v !== 'undefined' && v !== null && v !== false;

export const merge = <T>(objects: ({ [name: string]: T } | undefined | false)[] | undefined): { [name: string]: T } =>
  (objects || []).filter(isNotFalsy).reduce((i, acc) => ({ ...acc, ...i }), {});

export const typeToField = (type: string) => lowerFirst(type);

export const getLabel = (s: string) => startCase(camelCase(s));

export const or =
  <T>(...predicates: ((field: T) => boolean)[]) =>
  (field: T) =>
    predicates.some((predicate) => predicate(field));

export const and =
  <T>(...predicates: ((field: T) => boolean)[]) =>
  (field: T) =>
    predicates.every((predicate) => predicate(field));

export const not =
  <T>(predicate: (field: T) => boolean) =>
  (field: T) =>
    !predicate(field);

export const isRootModel = (model: EntityModel) => model.root;

export const isEntityModel = (model: Model): model is EntityModel => model instanceof EntityModel;

export const isEnumModel = (model: Model): model is EnumModel => model instanceof EnumModel;

export const isRawEnumModel = (model: Model): model is RawEnumModel => model instanceof RawEnumModel;

export const isScalarModel = (model: Model): model is ScalarModel => model instanceof ScalarModel;

export const isObjectModel = (model: Model): model is ObjectModel => model instanceof ObjectModel;

export const isInputModel = (model: Model): model is InputModel => model instanceof InputModel;

export const isInterfaceModel = (model: Model): model is InterfaceModel => model instanceof InterfaceModel;

export const isCreatableModel = (model: EntityModel) => model.creatable && model.fields.some(isCreatableField);

export const isUpdatableModel = (model: EntityModel) => model.updatable && model.fields.some(isUpdatableField);

export const isCreatableField = (field: EntityField) => !field.inherited && !!field.creatable;

export const isUpdatableField = (field: EntityField) => !field.inherited && !!field.updatable;

export const modelNeedsTable = (model: EntityModel) => model.fields.some((field) => !field.inherited);

export const hasName = (name: string) => (field: EntityField) => field.name == name;

export const isPrimitive = (field: EntityField): field is PrimitiveField =>
  field.kind === undefined || field.kind === 'primitive';

export const isEnum = (field: EntityField): field is EnumField => field.kind === 'enum';

export const isRelation = (field: EntityField): field is RelationField => field.kind === 'relation';

export const isInherited = (field: EntityField) => field.inherited;

export const isInTable = (field: EntityField) => field.name === 'id' || !field.inherited;

export const isToOneRelation = (field: EntityField): field is RelationField => isRelation(field) && !!field.toOne;

export const isQueriableField = ({ queriable }: EntityField) => queriable !== false;

export const isCustomField = (field: EntityField): field is CustomField => field.kind === 'custom';

export const isVisible = ({ hidden }: EntityField) => hidden !== true;

export const isSimpleField = and(not(isRelation), not(isCustomField));

export const isUpdatable = ({ updatable }: EntityField) => !!updatable;

export const isCreatable = ({ creatable }: EntityField) => !!creatable;

export const isQueriableBy = (role: string) => (field: EntityField) =>
  field.queriable !== false &&
  (field.queriable === undefined ||
    field.queriable === true ||
    !field.queriable.roles ||
    field.queriable.roles.includes(role));

export const isUpdatableBy = (role: string) => (field: EntityField) =>
  field.updatable && (field.updatable === true || !field.updatable.roles || field.updatable.roles.includes(role));

export const isCreatableBy = (role: string) => (field: EntityField) =>
  field.creatable && (field.creatable === true || !field.creatable.roles || field.creatable.roles.includes(role));

export const retry = async <T>(cb: () => Promise<T>, condition: (e: any) => boolean) => {
  try {
    return await cb();
  } catch (e) {
    if (condition(e)) {
      return await cb();
    } else {
      throw e;
    }
  }
};
