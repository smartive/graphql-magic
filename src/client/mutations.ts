import upperCase from 'lodash/upperCase';
import { isEntityModel } from '..';
import { RawModels } from '../models/models';

const constantCase = (str: string) => upperCase(str).replace(/ /g, '_');

export const generateMutations = (models: RawModels) => {
  const parts: string[] = [];
  for (const { name, creatable, updatable, deletable } of models.filter(isEntityModel)) {
    if (creatable) {
      parts.push(
        `export const CREATE_${constantCase(
          name
        )} = gql\`\n  mutation Create${name}Mutation($data: Create${name}!) {\n    create${name}(data: $data) { id }\n  }\n\`;`
      );
    }

    if (updatable) {
      parts.push(
        `export const UPDATE_${constantCase(
          name
        )} = gql\`\n  mutation Update${name}Mutation($id: ID!, $data: Update${name}!) {\n    update${name}(where: { id: $id }, data: $data) { id }\n  }\n\`;`
      );
    }

    if (deletable) {
      parts.push(
        `export const DELETE_${constantCase(
          name
        )} = gql\`\n  mutation Delete${name}Mutation($id: ID!) {\n    delete${name}(where: { id: $id })\n  }\n\`;`
      );
    }
  }

  return `import { gql } from "@smartive/graphql-magic";\n\n${parts.join('\n\n')}`;
};
