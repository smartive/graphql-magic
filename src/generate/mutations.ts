import { upperCase } from 'lodash';
import { Models } from '../models';

const constantCase = (str: string) => upperCase(str).replace(/ /g, '_');

export const generateMutations = (models: Models) => {
  const parts: string[] = [];
  for (const { name, creatable, updatable, deletable } of models) {
    if (creatable) {
      parts.push(
        `export const CREATE_${constantCase(
          name
        )} = gql\`\n  mutation Create${name}Mutation($data: Create${name}!) {\n    create${name}(data: $data) { id }\n  }\n\`;`
      );
    }

    if (updatable) {
      parts.push(
        `export const UPDATE_${constantCase(
          name
        )} = gql\`\n  mutation Update${name}Mutation($id: ID!, $data: Update${name}!) {\n    update${name}(where: { id: $id }, data: $data) { id }\n  }\n\`;`
      );
    }

    if (deletable) {
      parts.push(
        `export const DELETE_${constantCase(
          name
        )} = gql\`\n  mutation Delete${name}Mutation($id: ID!) {\n    delete${name}(where: { id: $id })\n  }\n\`;`
      );
    }
  }

  return `import { gql } from "../../../lib/graphql-magic/gql";\n\n${parts.join('\n\n')}`;
};

export const generateHooks = (models: Models) => {
  const hooks: string[] = [];

  const indexImports: string[] = [];
  const mutationImports: string[] = [];

  for (const { name, creatable, updatable, deletable } of models) {
    if (!(creatable || updatable || deletable)) {
      continue;
    }

    if (creatable) {
      indexImports.push(`Create${name}Mutation`);
      mutationImports.push(`CREATE_${constantCase(name)}`);
      hooks.push(
        `export const useCreate${name} = () => useMutation<Create${name}Mutation.mutation, Create${name}Mutation.Variables>(CREATE_${constantCase(
          name
        )});`
      );
    }

    if (updatable) {
      indexImports.push(`Update${name}Mutation`);
      mutationImports.push(`UPDATE_${constantCase(name)}`);
      hooks.push(
        `export const useUpdate${name} = () => useMutation<Update${name}Mutation.mutation, Update${name}Mutation.Variables>(UPDATE_${constantCase(
          name
        )});`
      );
    }

    if (deletable) {
      indexImports.push(`Delete${name}Mutation`);
      mutationImports.push(`DELETE_${constantCase(name)}`);
      hooks.push(
        `export const useDelete${name} = () => useMutation<Delete${name}Mutation.mutation, Delete${name}Mutation.Variables>(DELETE_${constantCase(
          name
        )});`
      );
    }
  }

  return `import {useMutation} from "graphql-hooks"

import {
    ${indexImports.sort().join(',\n    ')}
} from "./index";

import {
    ${mutationImports.sort().join(',\n    ')}
} from "./mutations";

${hooks.sort().join('\n\n')}\n`;
};
