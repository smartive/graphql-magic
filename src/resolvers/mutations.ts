import { GraphQLResolveInfo } from 'graphql';
import { v4 as uuid } from 'uuid';
import { Context } from '../context';
import { ForbiddenError, GraphQLError } from '../errors';
import { EntityField, EntityModel } from '../models/models';
import { Entity, MutationContext, Trigger } from '../models/mutation-hook';
import { get, isPrimitive, it, typeToField } from '../models/utils';
import { applyPermissions, checkCanWrite, getEntityToMutate } from '../permissions/check';
import { anyDateToLuxon } from '../utils';
import { resolve } from './resolver';
import { AliasGenerator, fetchDisplay, getTechnicalDisplay } from './utils';

export const mutationResolver = async (_parent: any, args: any, partialCtx: Context, info: GraphQLResolveInfo) => {
  return await partialCtx.knex.transaction(async (knex) => {
    const [, mutation, modelName] = it(info.fieldName.match(/^(create|update|delete|restore)(.+)$/));
    const ctx = { ...partialCtx, knex, info, aliases: new AliasGenerator() };
    switch (mutation) {
      case 'create': {
        const id = await createEntity(modelName, args.data, ctx, 'mutation');

        return await resolve(ctx, id);
      }
      case 'update': {
        const id = args.where.id;

        await updateEntity(modelName, id, args.data, ctx, 'mutation');

        return await resolve(ctx, id);
      }
      case 'delete': {
        const id = args.where.id;

        await deleteEntity(modelName, id, ctx, {
          dryRun: args.dryRun,
          trigger: 'mutation',
        });

        return id;
      }
      case 'restore': {
        const id = args.where.id;

        await restoreEntity(modelName, id, ctx, 'mutation');

        return id;
      }
    }
  });
};

export const createEntity = async (
  modelName: string,
  input: Entity,
  ctx: MutationContext,
  trigger: Trigger = 'direct-call',
) => {
  const model = ctx.models.getModel(modelName, 'entity');
  const normalizedInput = { ...input };
  if (!normalizedInput.id) {
    normalizedInput.id = uuid();
  }
  const id = normalizedInput.id as string;
  if (!normalizedInput.createdAt) {
    normalizedInput.createdAt = ctx.now;
  }
  if (!normalizedInput.createdById) {
    normalizedInput.createdById = ctx.user?.id;
  }
  if (model.parent) {
    normalizedInput.type = model.name;
  }
  sanitize(ctx, model, normalizedInput);

  await checkCanWrite(ctx, model, normalizedInput, 'CREATE');
  await ctx.handleUploads?.(normalizedInput);

  await ctx.mutationHook?.({
    model,
    action: 'create',
    trigger,
    when: 'before',
    data: { prev: {}, input, normalizedInput, next: normalizedInput },
    ctx,
  });

  if (model.parent) {
    const rootInput = {};
    const childInput = { id };
    for (const field of model.fields) {
      const columnName = field.kind === 'relation' ? `${field.name}Id` : field.name;
      if (columnName in normalizedInput) {
        if (field.inherited) {
          rootInput[columnName] = normalizedInput[columnName];
        } else {
          childInput[columnName] = normalizedInput[columnName];
        }
      }
    }
    await ctx.knex(model.parent).insert(rootInput);
    await ctx.knex(model.name).insert(childInput);
  } else {
    await ctx.knex(model.name).insert(normalizedInput);
  }
  await createRevision(model, normalizedInput, ctx);
  await ctx.mutationHook?.({
    model,
    action: 'create',
    trigger,
    when: 'after',
    data: { prev: {}, input, normalizedInput, next: normalizedInput },
    ctx,
  });

  return normalizedInput.id as string;
};

export const updateEntities = async (
  modelName: string,
  where: Record<string, unknown>,
  updateFields: Entity,
  ctx: MutationContext,
) => {
  const entities = await ctx.knex(modelName).where(where).select('id');
  for (const entity of entities) {
    await updateEntity(modelName, entity.id, updateFields, ctx);
  }
};

export const updateEntity = async (
  modelName: string,
  id: string,
  input: Entity,
  ctx: MutationContext,
  trigger: Trigger = 'direct-call',
) => {
  const model = ctx.models.getModel(modelName, 'entity');
  const normalizedInput = { ...input };

  sanitize(ctx, model, normalizedInput);

  const currentEntity = await getEntityToMutate(ctx, model, { id }, 'UPDATE');

  // Remove data that wouldn't mutate given that it's irrelevant for permissions
  for (const key of Object.keys(normalizedInput)) {
    if (normalizedInput[key] === currentEntity[key]) {
      delete normalizedInput[key];
    }
  }

  if (Object.keys(normalizedInput).length > 0) {
    await checkCanWrite(ctx, model, normalizedInput, 'UPDATE');
    await ctx.handleUploads?.(normalizedInput);

    await ctx.mutationHook?.({
      model,
      action: 'update',
      trigger,
      when: 'before',
      data: { prev: currentEntity, input, normalizedInput, next: { ...currentEntity, ...normalizedInput } },
      ctx,
    });
    await doUpdate(model, currentEntity, normalizedInput, ctx);
    await ctx.mutationHook?.({
      model,
      action: 'update',
      trigger,
      when: 'after',
      data: { prev: currentEntity, input, normalizedInput, next: { ...currentEntity, ...normalizedInput } },
      ctx,
    });
  }
};

type Callbacks = (() => Promise<void>)[];

export const deleteEntities = async (modelName: string, where: Record<string, unknown>, ctx: MutationContext) => {
  const entities = await ctx.knex(modelName).where(where).select('id');
  for (const entity of entities) {
    await deleteEntity(modelName, entity.id, ctx, {
      trigger: 'direct-call',
    });
  }
};

export const deleteEntity = async (
  modelName: string,
  id: string,
  ctx: MutationContext,
  {
    dryRun = false,
    trigger = 'direct-call',
  }: {
    dryRun?: boolean;
    trigger?: Trigger;
  } = {},
) => {
  const model = ctx.models.getModel(modelName, 'entity');
  const rootModel = model.rootModel;
  const entity = await getEntityToMutate(ctx, rootModel, { id }, 'DELETE');

  if (entity.deleted) {
    throw new ForbiddenError(`${getTechnicalDisplay(model, entity)} is already deleted.`);
  }

  const toDelete: Record<string, Record<string, string>> = {};
  const toUnlink: Record<
    string,
    Record<
      string,
      {
        display: string;
        fields: string[];
      }
    >
  > = {};
  const restricted: Record<
    string,
    Record<
      string,
      {
        display: string;
        fields: string[];
      }
    >
  > = {};

  const beforeHooks: Callbacks = [];
  const mutations: Callbacks = [];
  const afterHooks: Callbacks = [];

  const mutationHook = ctx.mutationHook;
  const deleteCascade = async (currentModel: EntityModel, currentEntity: Entity, currentTrigger: Trigger) => {
    if (!(currentModel.name in toDelete)) {
      toDelete[currentModel.name] = {};
    }
    if ((currentEntity.id as string) in toDelete[currentModel.name]) {
      return;
    }
    toDelete[currentModel.name][currentEntity.id as string] = await fetchDisplay(ctx.knex, currentModel, currentEntity);

    if (!dryRun) {
      const normalizedInput = {
        deleted: true,
        deletedAt: ctx.now,
        deletedById: ctx.user?.id,
        deleteRootType: rootModel.name,
        deleteRootId: entity.id,
      };
      if (mutationHook) {
        beforeHooks.push(async () => {
          await mutationHook({
            model: currentModel,
            action: 'delete',
            trigger: currentTrigger,
            when: 'before',
            data: { prev: currentEntity, input: {}, normalizedInput, next: { ...currentEntity, ...normalizedInput } },
            ctx,
          });
        });
      }
      mutations.push(async () => {
        await doUpdate(currentModel, currentEntity, normalizedInput, ctx);
      });
      if (mutationHook) {
        afterHooks.push(async () => {
          await mutationHook({
            model: currentModel,
            action: 'delete',
            trigger: currentTrigger,
            when: 'after',
            data: { prev: currentEntity, input: {}, normalizedInput, next: { ...currentEntity, ...normalizedInput } },
            ctx,
          });
        });
      }
    }

    for (const {
      targetModel: descendantModel,
      field: { name, foreignKey, onDelete },
    } of currentModel.reverseRelations.filter((reverseRelation) => !reverseRelation.field.inherited)) {
      const query = ctx.knex(descendantModel.name).where({ [foreignKey]: currentEntity.id, deleted: false });
      const descendants = await query;
      if (descendants.length) {
        switch (onDelete) {
          case 'set-null':
            for (const descendant of descendants) {
              if (dryRun) {
                if (!toUnlink[descendantModel.name]) {
                  toUnlink[descendantModel.name] = {};
                }
                if (!toUnlink[descendantModel.name][descendant.id]) {
                  toUnlink[descendantModel.name][descendant.id] = {
                    display: await fetchDisplay(ctx.knex, descendantModel, descendant),
                    fields: [],
                  };
                }
                toUnlink[descendantModel.name][descendant.id].fields.push(name);
              } else {
                const normalizedInput = { [`${name}Id`]: null };
                if (mutationHook) {
                  beforeHooks.push(async () => {
                    await mutationHook({
                      model: descendantModel,
                      action: 'update',
                      trigger: 'set-null',
                      when: 'before',
                      data: { prev: descendant, input: {}, normalizedInput, next: { ...descendant, ...normalizedInput } },
                      ctx,
                    });
                  });
                }
                mutations.push(async () => {
                  await doUpdate(descendantModel, descendant, normalizedInput, ctx);
                });
                if (mutationHook) {
                  afterHooks.push(async () => {
                    await mutationHook({
                      model: descendantModel,
                      action: 'update',
                      trigger: 'set-null',
                      when: 'after',
                      data: { prev: descendant, input: {}, normalizedInput, next: { ...descendant, ...normalizedInput } },
                      ctx,
                    });
                  });
                }
              }
            }
            break;
          case 'restrict':
            if (dryRun) {
              if (!restricted[descendantModel.name]) {
                restricted[descendantModel.name] = {};
              }
              for (const descendant of descendants) {
                if (!restricted[descendantModel.name][descendant.id]) {
                  restricted[descendantModel.name][descendant.id] = {
                    display: await fetchDisplay(ctx.knex, descendantModel, descendant),
                    fields: [name],
                  };
                }
                restricted[descendantModel.name][descendant.id].fields.push(name);
              }
            } else {
              throw new ForbiddenError(
                `${getTechnicalDisplay(model, entity)} cannot be deleted because it has ${getTechnicalDisplay(descendantModel, descendants[0])}${descendants.length > 1 ? ` (among others)` : ''}.`,
              );
            }
            break;
          case 'cascade':
          default: {
            if (!descendantModel.deletable) {
              throw new ForbiddenError(
                `${getTechnicalDisplay(model, entity)} depends on ${getTechnicalDisplay(descendantModel, descendants[0])}${descendants.length > 1 ? ` (among others)` : ''} which cannot be deleted.`,
              );
            }
            applyPermissions(ctx, descendantModel.name, descendantModel.name, query, 'DELETE');
            const deletableDescendants = await query;
            const notDeletableDescendants = descendants.filter(
              (descendant) => !deletableDescendants.some((d) => d.id === descendant.id),
            );
            if (notDeletableDescendants.length) {
              throw new ForbiddenError(
                `${getTechnicalDisplay(model, entity)} depends on ${descendantModel.labelPlural} which you have no permissions to delete.`,
              );
            }
            for (const descendant of descendants) {
              await deleteCascade(descendantModel, descendant, 'cascade');
            }
            break;
          }
        }
      }
    }
  };

  await deleteCascade(rootModel, entity, trigger);

  for (const callback of [...beforeHooks, ...mutations, ...afterHooks]) {
    await callback();
  }

  if (dryRun) {
    throw new GraphQLError(`Delete dry run:`, {
      code: 'DELETE_DRY_RUN',
      toDelete,
      toUnlink,
      restricted,
    });
  }
};

export const restoreEntity = async (
  modelName: string,
  id: string,
  ctx: MutationContext,
  trigger: Trigger = 'direct-call',
) => {
  const model = ctx.models.getModel(modelName, 'entity');
  const rootModel = model.rootModel;

  const entity = await getEntityToMutate(ctx, rootModel, { id }, 'RESTORE');

  if (!entity.deleted) {
    throw new ForbiddenError(`${getTechnicalDisplay(model, entity)} is not deleted.`);
  }

  if (entity.deleteRootId) {
    if (!(entity.deleteRootType === rootModel.name && entity.deleteRootId === entity.id)) {
      throw new ForbiddenError(
        `Can't restore ${getTechnicalDisplay(model, entity)} directly. To restore it, restore ${entity.deleteRootType} ${entity.deleteRootId}.`,
      );
    }
  }

  const toRestore: Record<string, Set<string>> = {};

  const beforeHooks: Callbacks = [];
  const mutations: Callbacks = [];
  const afterHooks: Callbacks = [];

  const restoreCascade = async (currentModel: EntityModel, currentEntity: Entity, currentTrigger: Trigger) => {
    if (entity.deleteRootId || currentEntity.deleteRootId) {
      if (!(currentEntity.deleteRootType === model.name && currentEntity.deleteRootId === entity.id)) {
        return;
      }

      // Legacy heuristic
    } else if (
      !anyDateToLuxon(currentEntity.deletedAt, ctx.timeZone)!.equals(anyDateToLuxon(entity.deletedAt, ctx.timeZone)!)
    ) {
      return;
    }

    if (!(currentModel.name in toRestore)) {
      toRestore[currentModel.name] = new Set();
    }
    toRestore[currentModel.name].add(currentEntity.id as string);

    const normalizedInput: Entity = {
      deleted: false,
      deletedAt: null,
      deletedById: null,
      deleteRootType: null,
      deleteRootId: null,
    };
    if (ctx.mutationHook) {
      beforeHooks.push(async () => {
        await ctx.mutationHook!({
          model: currentModel,
          action: 'restore',
          trigger: currentTrigger,
          when: 'before',
          data: { prev: currentEntity, input: {}, normalizedInput, next: { ...currentEntity, ...normalizedInput } },
          ctx,
        });
      });
    }
    mutations.push(async () => {
      for (const relation of currentModel.relations) {
        const parentId = currentEntity[relation.field.foreignKey] as string | undefined;
        if (!parentId) {
          continue;
        }
        if (toRestore[relation.targetModel.name]?.has(parentId)) {
          continue;
        }
        const parent = await ctx.knex(relation.targetModel.name).where({ id: parentId }).first();
        if (parent?.deleted) {
          throw new ForbiddenError(
            `Can't restore ${getTechnicalDisplay(model, entity)} because it depends on deleted ${relation.targetModel.name} ${parentId}.`,
          );
        }
      }

      await doUpdate(currentModel, currentEntity, normalizedInput, ctx);
    });
    if (ctx.mutationHook) {
      afterHooks.push(async () => {
        await ctx.mutationHook!({
          model: currentModel,
          action: 'restore',
          trigger: currentTrigger,
          when: 'after',
          data: { prev: currentEntity, input: {}, normalizedInput, next: { ...currentEntity, ...normalizedInput } },
          ctx,
        });
      });
    }

    for (const {
      targetModel: descendantModel,
      field: { foreignKey },
    } of currentModel.reverseRelations
      .filter((reverseRelation) => !reverseRelation.field.inherited)
      .filter(({ targetModel: { deletable } }) => deletable)) {
      const query = ctx.knex(descendantModel.name).where({ [foreignKey]: currentEntity.id, deleted: true });
      if (currentEntity.deleteRootId) {
        query.where({ deleteRootType: currentEntity.deleteRootType, deleteRootId: currentEntity.deleteRootId });
      } else {
        // Legacy heuristic
        query.where({ deletedAt: currentEntity.deletedAt });
      }
      const descendantsToRestore = await query;
      applyPermissions(ctx, descendantModel.name, descendantModel.name, query, 'RESTORE');
      const restorableDescendants = await query;
      const notRestorableDescendants = descendantsToRestore.filter(
        (descendant) => !restorableDescendants.some((d) => d.id === descendant.id),
      );
      if (notRestorableDescendants.length) {
        throw new ForbiddenError(
          `${getTechnicalDisplay(currentModel, currentEntity)} depends on ${descendantModel.labelPlural} which you have no permissions to restore.`,
        );
      }
      for (const descendant of descendantsToRestore) {
        await restoreCascade(descendantModel, descendant, 'cascade');
      }
    }
  };

  await restoreCascade(rootModel, entity, trigger);

  for (const callback of [...beforeHooks, ...mutations, ...afterHooks]) {
    await callback();
  }
};

export const createRevision = async (model: EntityModel, data: Entity, ctx: MutationContext) => {
  if (model.updatable) {
    const revisionId = uuid();
    const rootRevisionData: Entity = {
      id: revisionId,
      [`${typeToField(model.parent || model.name)}Id`]: data.id,
      createdAt: ctx.now,
      createdById: ctx.user?.id,
    };

    if (model.deletable) {
      rootRevisionData.deleted = data.deleted || false;
    }
    const childRevisionData = { id: revisionId };

    for (const field of model.fields.filter(({ updatable }) => updatable)) {
      const col = field.kind === 'relation' ? `${field.name}Id` : field.name;
      let value;
      if (field.nonNull && (!(col in data) || col === undefined || col === null)) {
        value = get(field, 'defaultValue');
      } else {
        value = data[col];
      }
      if (!model.parent || field.inherited) {
        rootRevisionData[col] = value;
      } else {
        childRevisionData[col] = value;
      }
    }

    if (model.parent) {
      await ctx.knex(`${model.parent}Revision`).insert(rootRevisionData);
      await ctx.knex(`${model.name}Revision`).insert(childRevisionData);
    } else {
      await ctx.knex(`${model.name}Revision`).insert(rootRevisionData);
    }
  }
};

const sanitize = (ctx: MutationContext, model: EntityModel, data: Entity) => {
  if (model.updatable) {
    if (!data.updatedAt) {
      data.updatedAt = ctx.now;
    }
    if (!data.updatedById) {
      data.updatedById = ctx.user?.id;
    }
  }

  for (const key of Object.keys(data)) {
    const field = model.fields.find(({ name }) => name === key);

    if (!field) {
      continue;
    }

    if (isEndOfDay(field) && data[key]) {
      data[key] = anyDateToLuxon(data[key], ctx.timeZone)!.endOf('day');
      continue;
    }

    if (isEndOfMonth(field) && data[key]) {
      data[key] = anyDateToLuxon(data[key], ctx.timeZone)!.endOf('month');
      continue;
    }

    if (field.list && field.kind === 'enum' && Array.isArray(data[key])) {
      data[key] = `{${(data[key] as string[]).join(',')}}`;
      continue;
    }
  }
};

const isEndOfDay = (field: EntityField) =>
  isPrimitive(field) && field.type === 'DateTime' && field?.endOfDay === true && field?.dateTimeType === 'date';

const isEndOfMonth = (field: EntityField) =>
  isPrimitive(field) && field.type === 'DateTime' && field?.endOfMonth === true && field?.dateTimeType === 'year_and_month';

const doUpdate = async (model: EntityModel, currentEntity: Entity, update: Entity, ctx: MutationContext) => {
  if (model.updatable) {
    if (!update.updatedAt) {
      update.updatedAt = ctx.now;
    }
    if (!update.updatedById) {
      update.updatedById = ctx.user?.id;
    }
  }
  if (model.parent) {
    const rootInput = {};
    const childInput = {};
    for (const field of model.fields) {
      const columnName = field.kind === 'relation' ? `${field.name}Id` : field.name;
      if (columnName in update) {
        if (field.inherited) {
          rootInput[columnName] = update[columnName];
        } else {
          childInput[columnName] = update[columnName];
        }
      }
    }
    if (Object.keys(rootInput).length) {
      await ctx.knex(model.parent).where({ id: currentEntity.id }).update(rootInput);
    }
    if (Object.keys(childInput).length) {
      await ctx.knex(model.name).where({ id: currentEntity.id }).update(childInput);
    }
  } else {
    await ctx.knex(model.name).where({ id: currentEntity.id }).update(update);
  }
  await createRevision(model, { ...currentEntity, ...update }, ctx);
};
