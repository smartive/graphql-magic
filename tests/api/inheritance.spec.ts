import { gql } from '../../src';
import { DELETE_REVIEW, RESTORE_REVIEW } from '../generated/client/mutations';
import { REVIEW_ID } from '../utils/database/seed';
import { withServer } from '../utils/server';

describe('inheritance', () => {
  describe('queries', () => {
    it('root type listQuery', async () => {
      await withServer(async (request) => {
        expect(
          await request(gql`
            query GetReactions {
              reactions {
                type
                content
                ... on Review {
                  rating
                }
              }
            }
          `)
        ).toMatchSnapshot();
      });
    });

    it('root type query', async () => {
      await withServer(async (request) => {
        expect(
          await request(gql`
            query GetReaction {
              reaction(where: { id: "${REVIEW_ID}" }) {
                type
                content
                ... on Review {
                  rating
                }
              }
            }
          `)
        ).toMatchSnapshot();
      });
    });

    it('sub type listQuery', async () => {
      await withServer(async (request) => {
        expect(
          await request(gql`
            query GetReviews {
              reviews {
                type
                content
                rating
              }
            }
          `)
        ).toMatchSnapshot();
      });
    });

    it('sub type query', async () => {
      await withServer(async (request) => {
        expect(
          await request(gql`
            query GetReview {
              review(where: { id: "${REVIEW_ID}" }) {
                type
                content
                rating
              }
            }
          `)
        ).toMatchSnapshot();
      });
    });

    // TODO
    // root type -> root type relation
    // root type -> sub type relation
    // sub type -> root type relation
    // sub type -> sub type relation
    // reverse relations...
    // basically all schema possibilities...
  });

  describe('mutations', () => {
    it('create', async () => {
      await withServer(async (request) => {
        expect(
          await request(
            gql`
              mutation CreateReview($data: CreateReview!) {
                createReview(data: $data) {
                  content
                  rating
                }
              }
            `,
            { data: { content: 'A review', rating: 5 } }
          )
        ).toMatchSnapshot();
      });
    });

    it('update', async () => {
      await withServer(async (request) => {
        expect(
          await request(
            gql`
              mutation UpdateReview($id: ID!, $data: UpdateReview!) {
                updateReview(where: { id: $id }, data: $data) {
                  content
                  rating
                }
              }
            `,
            { id: REVIEW_ID, data: { content: 'A review', rating: 5 } }
          )
        ).toMatchSnapshot();
      });
    });

    it('delete and restore', async () => {
      await withServer(async (request) => {
        expect(await request(DELETE_REVIEW, { id: REVIEW_ID })).toMatchSnapshot();

        expect(await request(RESTORE_REVIEW, { id: REVIEW_ID })).toMatchSnapshot();
      });
    });
  });
});
