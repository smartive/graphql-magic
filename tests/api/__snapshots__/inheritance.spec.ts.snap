// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inheritance mutations create 1`] = `
{
  "createReview": {
    "content": "A review",
    "rating": 5,
  },
}
`;

exports[`inheritance mutations delete and restore 1`] = `
{
  "mutated": "817c55de-2f77-4159-bd44-9837d868f889",
}
`;

exports[`inheritance mutations delete and restore 2`] = `
{
  "mutated": "817c55de-2f77-4159-bd44-9837d868f889",
}
`;

exports[`inheritance mutations update 1`] = `
{
  "updateReview": {
    "content": "A review",
    "rating": 5,
  },
}
`;

exports[`inheritance queries root type listQuery 1`] = `
{
  "reactions": [
    {
      "content": "This is a review with a rating",
      "rating": 5,
      "type": "Review",
    },
    {
      "content": "I do not know but here is the answer.",
      "type": "Answer",
    },
    {
      "content": "What is the question?",
      "type": "Question",
    },
  ],
}
`;

exports[`inheritance queries root type query 1`] = `
{
  "reaction": {
    "content": "This is a review with a rating",
    "rating": 5,
    "type": "Review",
  },
}
`;

exports[`inheritance queries sub type listQuery 1`] = `
{
  "reviews": [
    {
      "content": "This is a review with a rating",
      "rating": 5,
      "type": "Review",
    },
  ],
}
`;

exports[`inheritance queries sub type query 1`] = `
{
  "review": {
    "content": "This is a review with a rating",
    "rating": 5,
    "type": "Review",
  },
}
`;
